vi-start-to-finish.py

```diff
@@ -0,0 +1,223 @@
+from microscopium import preprocess as pre
+from microscopium._util import generate_spiral
+from microscopium import features as feat
+
+
+ROOT = '/scratch/su62/petermac/data'
+os.chdir(ROOT)
+OUT = os.path.join(ROOT, 'out')
+os.makedirs(OUT, exist_ok=True)
+
+
+def ftime(seconds):
+    string = ''
+    units = ['d', 'h', 'm', 's']
+    values = [3600 * 24, 3600, 60, 1]
+    for unit, value in zip(units, values):
+        amount = seconds // value
+        if amount > 0:
+            string += f'{amount}{unit}'
+        seconds = seconds % value
+    return string
+
+# config in /home/jnun0003/.config/dask/jobqueue.yaml
+cluster = Cluster(walltime='06:00:00')
+cluster.adapt(minimum=2, maximum=16, target_duration='1d')
+
+############### illumination fields #####################
+t0 = time.time()
+
+exp = (r'(?P<dir>.*)/(?P<plate>mfgtmp_\d*)_'
+       r'(?P<well>[A-P]\d\d)f(?P<field>\d\d)d(?P<channel>\d).TIF')
+
+def find_background(key):
+    print(f'finding background for {key}')
+    plate, ch = key
+    fns = glob.glob(f'{plate}/{plate}_*d{ch}.TIF')
+    illum = pre.find_background_illumination(fns, radius=151)
+    return illum
+
+# set up dask Futures to be computed on cluster (client)
+dirs = filter(lambda x: x.startswith('mfgtmp'), os.listdir('.'))
+keys = list(itertools.product(dirs, ['0', '1']))
+client = Client(cluster)
+illum_fields = {k: client.submit(find_background, k) for k in keys}
+
+# now gather the results
+for k in tqdm(illum_fields, 'illum'):
+    future = illum_fields[k]
+    illum_fields[k] = future.result()
+    future.cancel()
+    im = illum_fields[k]
+    fn = 'illum-' + str.join('-', k) + '.png'
+    io.imsave(os.path.join(OUT, fn), im)
+
+t1 = time.time()
+print(f'illumination estimated in {ftime(t1 - t0)}')
+
+############### illumination correction #####################
+# set up dask Futures to be computed on cluster (client)
+def correct_illumination(tup):
+    key, illum_field = tup
+    print(f'fixing background for {key}')
+    plate, ch = key
+    fns = glob.glob(f'{plate}/{plate}_*d{ch}.TIF')
+    corrected = pre.correct_multiimage_illumination(fns, illum_field,
+                                                    stretch_quantile=0.01)
+    for fn, image in zip(fns, corrected):
+        fnout = fn[:-4] + '.illum.png'
+        io.imsave(fnout, image)
+    return key
+
+
+futures = client.map(correct_illumination, illum_fields.items())
+
+for fut in tqdm(as_completed(futures), desc='illum-corr', total=len(keys)):
+    r = fut.result()
+    fut.cancel()
+
+t2 = time.time()
+print(f'illumination corrected in {ftime(t2 - t1)}')
+
+############### montaging #####################
+
+# these are much shorter tasks, so we start a new cluster:
+cluster.close()
+cluster = Cluster(walltime='01:00:00', cores=1, memory='4gb', job_cpu=1, job_mem='4gb')
+cluster.adapt(minimum=0, maximum=48, target_duration='4h')
+
+
+intermediate_images = sorted(glob.glob('*/*.illum.png'))
+exp = (r'(?P<dir>.*)/(?P<plate>mfgtmp_\d*)_'
```

Comment:
Didn't you just define `exp` above?

Reply:
Yes, this is the real definition, have moved to the top.